### PR TITLE
FIX: support URL to any forge

### LINKED ISSUE #92: repo should work for non GitHub urls
Currently we cannot parse arbitrary url for repo...  See #91

### PR DESCRIPTION
**Pull request by tfardet**
Fixes #92 by accepting any URL in the ``repo`` field.

### CODE REVIEW

File: python/build.py [hunk 1]
```diff
@@ -27,11 +27,18 @@
...
+        try:
+            # try outdated GitHub user/repo_name format
+            _, package['repo_name'] = package['repo'].split('/')
+            # set proper repository URL
+            package['repo'] = f"https://github.com/{package['repo']}"
+        except ValueError:
+            warnings.warn(f'Package.repo is not in correct format: {package}')
+            continue
+
+    package.setdefault('repo_name', package['name'])
```
Comment (timhoffm) [inline python/build.py:41]: I'm slightly worried about this. `name` is the name for display and may not be a legal reponame (e.g. "Farrow&Ball Matplotlib"). This is brittle. It works as long as we have a `repo_name` (i.e. the old GitHub user/repo_name format), but may fail for other `repo` urls if the do not specify `repo_name` explicitly.

`repo_name` is currently used for `conda_package`, `pypi_name`, `rtd_name` and `site`. Overall, I think we should move away from assumtions and towards simple explicit declarations. But that may be a bit much for this PR.

There are two ways forward:
- Get rid of all the inference and assumtions: This means somebody has to go through all the config files, and write down things explicitly.
- Only use `package['name']` if it's a reasonably simple identifier, e.g. (`re.match(r'^[\w-]+$', package['name'])`. But that means 'repo_name' is not always available and would need to be handled on the usage sites.

I'd be ok, to accept this PR as is if we open a follow-up issue to go down one these roads.

**Comment by tfardet** [comment]
I agree that current state is suboptimal, I could add a regex check in this PR and raise an error if regex fails and repo is a URL until we fix this?

**Comment by QuLogic** [comment]
You can add the regex to the schema, which should be caught by pre-commit.

**Comment by timhoffm** [comment]
> You can add the regex to the schema, which should be caught by pre-commit.

This does not help here. The name parameter should still be free text. But we can only try to infer the repo name from it if it's format is more restricted.

**Review by timhoffm — APPROVED** [review]

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -4,9 +4,12 @@
...
+at the minimum you need to include:
 
-Other useful fields are `site:`, `pypi_name`, `conda_package`, (if different from the
-github name), and `conda_channel` if not *conda-forge*.
+* `name:`, the package name
+* `repo:`, the URL to the package repository
+* `section:`, the type of package (pick one of the sections on the [3rd-party page](https://matplotlib.org/mpl-third-party/))
+* a *very* short `description:` field.
+
+Other useful fields are `site:`, `pypi_name`, `conda_package`, (if different from the name), and `conda_channel` if not *conda-forge*.
```
Comment (QuLogic) [inline .github/pull_request_template.md:14]: Should remove the colons, I think, or at least be consistent:
```suggestion
* `name`, the package name
* `repo`, the URL to the package repository
* `section`, the type of package (pick one of the sections on the [3rd-party page](https://matplotlib.org/mpl-third-party/))
* a *very* short `description` field.

Other useful fields are `site`, `pypi_name`, `conda_package` (if different from the name), and `conda_channel` if not *conda-forge*.
```

**Review by QuLogic — APPROVED** [review]

**Comment by QuLogic** [comment]
Thanks @tfardet! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

**Pull request merged**